README.md
A Flutter and Dart plugin allowing you to convert the JSON format used by firestore to represent its documents or
collections to the conventional JSON data format

## Motivation

As a flutter developer, developing for Windows and Linux platforms, we noticed that none of the packages (
cloud_firestore, etc.) provided by Firebase supported desktop platforms (except MacOs). Nevertheless, there is still the
solution with the REST APIs provided by Firebase. But a problem that arises when working with Firebase REST API mainly
with Cloud Firestore, this problem is that the data provided by Cloud Firestore is in a rather "special" JSON format
which can slow down your development time if you try to take some more time to analyze the data and apply a conversion
yourself.
It is with this in mind that we have created this package to convert documents and/or collections that are in the "
firestore format" to a "standard" json format.

## Use case

This package has been designed to work with Firebase's Cloud Firestore REST API.
If you are developing a flutter application on Windows, Linux or a Dart console application (except macOS) you can use
this package to help you convert data from a document or a Firestore collection. Because the currently available
cloud_firestore plugin does not support these platforms.

- If you are developing for a platform like iOS, macOS, Android or Web we recommend you to use the official
  cloud_firestore package which does a great job.
- You can also use this plugin to receive, send and convert your data from a Firestore document or collection via the
  Cloud Firestore REST API.

## Features

Convert/encode a json data schema into a format understandable by the cloud firestore rest api and vice versa

#### Supported values

Regarding the conversion of a json to the format read by the cloud firestore rest api and vice versa

All values are supported:

- stringValue (String)
- integerValue (int)
- doubleValue (double)
- booleanValue (bool)
- mapValue (Map)
- arrayValue (List)
- timestampValue (DateTime)
- geoPointValue (Map)
- referenceValue (String)
- nullValue (null)

## Getting started

Installation

Add this line to pubspec.yaml

```yaml
dependencies:
firestore_api_parser: <latest_version> #add this
```

Import package

```dart
import 'package:firestore_api_parser/firestore_api_parser.dart';
```

## Usage

A simple usage example. For a more concrete and detailed example, please consult the [example file][example].

```dart
void main() {
  final names = {'lastname': 'BOND', 'firstname': 'James', 'initial': 'JB'};

  final bond = Agent(
    names: names,
    missions: [
      {'2022': 'No Time To Die'},
      {'2015': 'Spectre'},
      {'2006': 'Casino Royale'},
    ],
    status: 'In service',
    round: 20.5,
    timestamp: DateTime.parse('2021-10-07T19:00:00Z'),
    storageRef: 'projects/my_great_project/databases/(default)/documents/USERS/abc123',
    nullable: null,
    coordinates: {'latitude': -64, 'longitude': -86},
  );

  // Will convert [bond] to the format json used by Firestore format documents
  final firestoreJsonDoc = FirestoreApiParser.convertToFirestoreDocument(json: bond.toJson());

  //using extension method
  // final firestoreJsonDoc = bond.toJson().toFirestoreJson;

  // Will convert [firestoreJsonDoc] to a "standard" JSON format representation
  final json = FirestoreApiParser.convertToStandardJson(doc: firestoreJsonDoc);

  //using extension method
  // final json =  firestoreJsonDoc.toStandardJson;

  // You can use this method bellow to parse & convert firestore collection to JSON
  // final document = FirestoreApiParser.parseCollection(collection: firestoreCollection);
}
```

## Additional information

- You can use extension methods to encode or decode data

```dart

final firestoreJsonDoc = bond
    .toJson()
    .toFirestoreJson; //encode data to firestore Json structure using extension methods

final json =  firestoreJsonDoc.toStandardJson; //decode data using extension methods
```

- To insert a **referenceValue**, the value must match this pattern `projects/{{my_great_project}}/databases/(default)
  /documents/` according to the firestore structure

i.e:  this will be recognized as a **referenceValue** by the package and cloud_firestore

```dart

final reference = 'projects/{{my_great_project}}/databases/(default)/documents/';
// also you can add collection or doc 
// projects/{{my_great_project}}/databases/(default)/documents/{{my_collection}}  //point to a collection
// projects/{{my_great_project}}/databases/(default)/documents/{{my_collection}}/{{docid}} //point to a document
```

- To insert a geoPointValue, you must create a map that contains only 2 keys **longitude** and **latitude** and
  their values

i.e:

```dart

final geoPointValue = {'longitude': -86, 'latitude': -64};
```

- To insert data as a **timestampValue**, you need to pass the date in _isoString_ format like **YYYY-MM-DDTHH:MM:SSZ**.
  If the conversion of the date to the firestore json format fails, the value will be inserted as a **stringValue**

- For a better result and to avoid bugs or unexpected behaviors, we recommend you to use
  the [json_serializable](https://pub.dev/packages/json_serializable) package (or any other JSON serialization tool) to
  serialize your data / object into JSON as follows:

i.e: json_serializable

```dart
import 'package:json_annotation/json_annotation.dart';

part 'agent.g.dart';

@JsonSerializable()
class Agent {
  final String storageRef;
  final DateTime timestamp;
  final String? nullable;
  final double round;
  final Map<String, num> coordinates;
  final String status;
  final Map<String, dynamic> names;
  final List missions;

  Agent({
    required this.coordinates,
    required this.storageRef,
    required this.timestamp,
    required this.round,
    required this.names,
    required this.missions,
    required this.status,
    this.nullable,

  });

  factory Agent.fromJson(Map<String, dynamic> data) => _$AgentFromJson(data);

  Map<String, dynamic> toJson() => _$AgentToJson(this);
}
```

## Features and bugs

For any suggestions or feature requests, please contact us at _tech.riverfortress@gmail.com_ .

For any bugs report, post your problem in the  [issue tracker][tracker] section.

Made with ❤

Stay safe!

[tracker]: https://github.com/stvndelucis/firestore_api_parser/issues

[example]: https://github.com/stvndelucis/firestore_api_parser/blob/main/example/example.dart


CHANGELOG.md
## 1.0.0

Initial version. A simple parser allowing you to convert a json that is in firebase cloud firestore format to the
conventional json format.
For the moment the API only allows to convert cloud firestore documents. An update to support the conversion of cloud
firestore collection is planned for the next updates of the package

## 1.0.1

Adding support for meta^1.3.0

## 1.0.2

Adding support these package :

- json_serializable: ^4.1.4
- json_annotation: ^4.0.1

## 1.0.3

In this update, we

- Support for parsing and converting cloud firestore collection to standard json format, by using **parseCollection**
  method
- have fixed some minor bugs
- Added support for parsing doubleValue. Sorry to have forgotten, we are sorry for the inconvenience caused

## 1.0.4

- Fix minor bugs about parsing of a cloud firestore arrayValue
- Deprecating toJson and toFirestoreJson methods

## 1.0.5

- Fixed the bug that unintentionally converted a **string** into a **timestamp**.
  Now, for a string to be recognized in **timestamp** format in Firestore, it must be in **Iso8601 utc** format (so it
  must end with Z) otherwise it will be stored in Firestore as **stringValue** if the conversion fails.
- Updated plugin example use
- Updated plugin description

## 1.0.6

- Updated README.md file

## 1.1.0

**BREAKING CHANGES!**

1. Improvement of the method `parseCollection()`. in order to meet the need to convert the documents returned by cloud
   firestore following a **structuredQuery**, we have made some modifications to the `parseCollection()` method so that
   it is able to convert the data from a structured query to standard json format.
   So we introduced the optional parameter `isStructuredQuery` which by default is `false`.
   When you perform structuredQuery with the cloud firestore REST API, you will only have to set the `isStructuredQuery`
   parameter to `true` so that it can parse and convert the data from the collection you provide.

2. Fixed the bug that prevented the conversion/parsing of documents from a Firestore collection obtained from the result
   of a structuredQuery

3. Suppression of depreciated methods `toFirestoreJson()`, `toJson()`

4. Updated README.md file

## 1.1.1

- intl package upgraded to ^0.18.0
- environnement constraint upgraded to '>=2.12.0 <4.0.0'

## 2.0.0

**BREAKING CHANGES!**

1. Parsing a structured query collection support was removed until we can implement a solid, viable solution... Coming
   soon :).
2. Adding extension methods `toFirestoreJson` and `toStandardJson` to encode and decode documents. Also
   added `parseCollection` extension method to decode a firestore collection.
3. Improved code structure to enhance maintainability.
4. Map and Array firestore now return empty values instead of throwing an exception when "fields" or "values" are not
   supplied.
5. Updated project dependencies.
6. Improved package tests.
7. Updated example and README file.

test/firestore_to_json_test.dart
import 'package:firestore_api_parser/firestore_api_parser.dart';
import 'package:test/test.dart';

void main() {
  group('Testing extractMapData function', () {
    test('Trying to parse a fake firestore map that not contains the "fields" field. Expecting to throw an Exception',
        () {
      final fakeFirestoreMap = {
        '2006': {'stringValue': 'Casino Royale'}
      };

      expect(() => decodeMapData(fakeFirestoreMap), throwsException);
    });

    test(
        'Trying to parse a good firestore map that contains "fields" field.\n'
        'Expecting to get parsed data to a json format', () {
      final goodFirestoreMap = {
        'fields': {
          '2006': {'stringValue': 'Casino Royale'}
        }
      };

      final expectedData = {'2006': 'Casino Royale'};

      final parsedData = decodeMapData(goodFirestoreMap);

      expect(parsedData['2006'].toString(), expectedData['2006']);
    });
  });

  group('Testing extractArrayData function', () {
    test('retrieving a array Value with that contains "values" fields', () {
      final arrayValue = {
        'values': [
          {'stringValue': 'Data 1'},
          {'stringValue': 'Data 2'},
          {'stringValue': 'Data 3'}
        ]
      };
      final parsedArray = decodeArrayData(arrayValue);

      expect(parsedArray.length, 3);
    });

    test('retrieving a array Value with that does not contains "values" fields', () {
      final arrayValue = {
        'data': [
          {'stringValue': 'Data 1'},
          {'stringValue': 'Data 2'},
          {'stringValue': 'Data 3'}
        ]
      };

      final parsedArray = decodeArrayData(arrayValue);
      expect(parsedArray.length, 0);
    });
  });

  group('Testing parseToJson function from firestore map', () {
    test('retrieving a String Value', () {
      final stringValue = {'stringValue': 'Steeven Delucis'};

      final parsedValue = parseToJson(stringValue);
      expect(parsedValue.runtimeType, String);
      expect(parsedValue, 'Steeven Delucis');
    });

    test('retrieving a int Value', () {
      final intValue = {'integerValue': 26};
      final parsedValue = parseToJson(intValue);

      expect(parsedValue.runtimeType, int);
      expect(parsedValue, 26);
    });

    test('retrieving a double Value', () {
      final doubleValue = {'integerValue': 26.5};
      final parsedValue = parseToJson(doubleValue);

      expect(parsedValue.runtimeType, double);
      expect(parsedValue, 26.5);
    });

    test('retrieving a timestamp (DateTime) Value', () {
      final timestampValue = {'timestampValue': '2021-10-07T19:00:00Z'};
      final parsedValue = parseToJson(timestampValue);

      final dateTime = DateTime.parse(parsedValue);

      expect(parsedValue.runtimeType, String);
      expect(parsedValue, '2021-10-07T19:00:00Z');

      expect(dateTime.runtimeType, DateTime);
    });

    test('retrieving a bool Value', () {
      final booleanValue = {'booleanValue': true};
      final parsedValue = parseToJson(booleanValue);

      expect(parsedValue.runtimeType, bool);
      expect(parsedValue, true);
    });

    test('retrieving a gepPoint Value', () {
      final booleanValue = {
        'geoPointValue': {'latitude': '-64', 'longitude': '-86'}
      };
      final parsedValue = parseToJson(booleanValue);

      expect(parsedValue['latitude'], '-64');
      expect(parsedValue['longitude'], '-86');
    });

    test('retrieving a reference Value', () {
      final referenceValue = {'referenceValue': 'projects/my_project/databases/(default)/documents/USERS/abc123'};
      final parsedValue = parseToJson(referenceValue);

      expect(parsedValue.runtimeType, String);
      expect(parsedValue, 'projects/my_project/databases/(default)/documents/USERS/abc123');
    });

    test('retrieving a null Value', () {
      final nullValue = {'nullValue': null};

      final parsedData = parseToJson(nullValue);
      expect(parsedData, null);
    });
  });
}


test/json_to_firestore_test.dart
import 'package:firestore_api_parser/firestore_api_parser.dart';
import 'package:test/test.dart';

void main() {
  group('encodeData method', () {
    //
    test('It should encode value as referenceValue because it match referenceValue pattern', () {
      final value = 'projects/my_great_project/databases/(default)/documents/USERS/abc123';

      final parsedValue = encodeData(value);

      final expected = {'referenceValue': 'projects/my_great_project/databases/(default)/documents/USERS/abc123'};

      expect(parsedValue, expected);
    });

    test('It should encode value as timestampValue because it match a valid date pattern', () {
      final value = '2021-10-07T19:00:00Z';

      final parsedValue = encodeData(value);

      final excepted = {'timestampValue': '2021-10-07T19:00:00Z'};

      expect(parsedValue, excepted);
    });

    test("It should encode value as a stringValue because it's a simple String", () {
      final value = 'Steeven Delucis';

      final parsedValue = encodeData(value);
      final expected = {'stringValue':'Steeven Delucis'};
      expect(parsedValue, expected);
    });

    test("It should encode value as booleanValue because it's a simple bool", () {
      final value = true;

      final parsedValue = encodeData(value);
      final expected = {'booleanValue': true};

      expect(parsedValue, expected);
    });

    test(
        "It should encode value as integerValue because it's a simple int."
        "value will be transformed to String in accordance with how cloud firestore encode int", () {
      final value = 26;

      final parsedValue = encodeData(value);
      final expected = {'integerValue': '26'};

      expect(parsedValue, expected);
    });

    test("It should encode value as doubleValue because it's a simple double", () {
      final doubleValue = 26.0;

      final parsedValue = encodeData(doubleValue);
      final expected = {'doubleValue': 26.0};

      expect(parsedValue, expected);
    });

    test("It should encode value as nullValue because value is null", () {
      final value = null;

      final parsedValue = encodeData(value);
      final expected = {'nullValue': null};

      expect(parsedValue, expected);
    });

    // And what if user only want to store it has a Map ? Thinkin' about it...
    test("It should encode value as geoPointValue because it match geoPointValue pattern", () {
      final value = {'latitude': '-64', 'longitude': '-84'};

      final parsedValue = encodeData(value);

      final expected = {
        'geoPointValue': {'longitude': '-84', 'latitude': '-64'}
      };

      expect(parsedValue, expected);
    });

    test('It should encode value as mapValue', () {
      final value = {
        'names': {'firstname': 'James', 'lastname': 'Bond'}
      };

      final parsedValue = encodeData(value);

      final expected = {
        'mapValue': {
          'fields': {
            'names': {
              'mapValue': {
                'fields': {
                  'firstname': {'stringValue': 'James'},
                  'lastname': {'stringValue': 'Bond'},
                }
              }
            }
          }
        }
      };

      expect(parsedValue, expected);
    });

    test('It should encode value as mapValue and dont create fields key when value is empty', () {
      final value = {};

      final parsedValue = encodeData(value);

      final expected = {'mapValue': {}};

      expect(parsedValue, expected);
    });

    test('It should encode array as arrayValue', () {
      final array = ['Billie Eillish', 667, 09.11];

      final encodedArray = encodeData(array);

      final expected = {
        'arrayValue': {
          'values': [
            {'stringValue': 'Billie Eillish'},
            {'integerValue': '667'},
            {'doubleValue': 09.11}
          ]
        }
      };

      expect(encodedArray, expected);
    });

    test('It should encode array as arrayValue and dont create a values key when value is empty', () {
      final array = [];

      final encodedArray = encodeData(array);

      final expected = {'arrayValue': {}};

      expect(encodedArray, expected);
    });

    test("It should throw an exception because value's type is not unsupported or cannot be determined", () {
      final Object value = Object();

      expect(() => encodeData(value), throwsException);
    });
  });
}


test/firestore_api_parser_test.dart
import 'package:firestore_api_parser/firestore_api_parser.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {


    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {

    });
  });
}


example/agent.g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'agent.dart';

// **************************************************************************
// JsonSerializableGenerator
// **************************************************************************

Agent _$AgentFromJson(Map<String, dynamic> json) => Agent(
      coordinates: Map<String, num>.from(json['coordinates'] as Map),
      storageRef: json['storageRef'] as String,
      timestamp: DateTime.parse(json['timestamp'] as String),
      round: (json['round'] as num).toDouble(),
      nullable: json['nullable'] as String?,
      names: json['names'] as Map<String, dynamic>,
      missions: json['missions'] as List<dynamic>,
      status: json['status'] as String,
    );

Map<String, dynamic> _$AgentToJson(Agent instance) => <String, dynamic>{
      'storageRef': instance.storageRef,
      'timestamp': instance.timestamp.toIso8601String(),
      'nullable': instance.nullable,
      'round': instance.round,
      'coordinates': instance.coordinates,
      'names': instance.names,
      'missions': instance.missions,
      'status': instance.status,
    };


lib/firestore_api_parser.dart
library firestore_api_parser;

import 'package:meta/meta.dart';

part 'src/firestore_to_json.dart';

part 'src/json_to_firestore.dart';

part 'src/firestore_collection_to_json.dart';

abstract class FirestoreApiParser {
  /// Parses and converts a [doc] that is in a format readable by cloud firestore to a standard json format
  ///
  /// returns a [Map] with converted data
  static Map<String, dynamic> convertToStandardJson({required Map<String, dynamic> doc}) =>
      toStandardJsonFormat(firestoreJson: doc);

  /// Analyze a standard [Map]/JSON which is in standard json format and
  /// returns [Map]/JSON with data converted to a readable format for Cloud Firestore
  static Map<String, dynamic> convertToFirestoreDocument({required Map<String, dynamic> json}) =>
      toFirestoreDocumentFormat(jsonMap: json);

  /// Analyze and convert a collection that is in json format used by cloud firestore to standard json format
  ///
  /// Returns a [Map] containing all the documents of the collection in standard json format
  static List<dynamic> parseCollection({required dynamic collection}) => collectionToJson(collection: collection);
}

extension FirestoreApiParserMapExtension on Map {
  Map<String, dynamic> get toFirestoreJson =>
      FirestoreApiParser.convertToFirestoreDocument(json: cast<String, dynamic>());

  Map<String, dynamic> get toStandardJson => FirestoreApiParser.convertToStandardJson(doc: cast<String, dynamic>());
}

extension FirestoreApiParserListExtension on List {
  List<dynamic> get parseCollection => FirestoreApiParser.parseCollection(collection: this);
}


lib/src/firestore_to_json.dart
part of firestore_api_parser;

@visibleForTesting
Map<String, dynamic> toStandardJsonFormat({required Map<String, dynamic> firestoreJson}) =>
    decodeMapData(firestoreJson);

@visibleForTesting
dynamic parseToJson(Map values) => values.map((key, value) {
      switch (key) {
        case 'stringValue':
        case 'integerValue':
        case 'doubleValue':
        case 'timestampValue':
        case 'booleanValue':
        case 'geoPointValue':
        case 'referenceValue':
          return MapEntry(key, value);

        case 'arrayValue':
          return MapEntry(key, decodeArrayData(value));

        case 'mapValue':
          return MapEntry(key, decodeMapData(value));

        case 'nullValue':
          return MapEntry(key, null);

        default:
          throw Exception('Cannot convert this value to a json readable format. '
              'That sound like this value type $key is not supported by firestore.\nReceived data $values');
      }
    });

@visibleForTesting
Map<String, dynamic> decodeMapData(Map<String, dynamic> mapData) {
  final fields = mapData['fields'] as Map<String, dynamic>?;

  if (fields == null) return const {};

  return fields.map((key, value) => MapEntry(key, parseToJson(value)));
}

@visibleForTesting
List<dynamic> decodeArrayData(Map arrayData) {
  if (arrayData['values'] == null) return [];

  final values = arrayData['values'] as List<dynamic>;

  return values.map((value) => parseToJson(value)).toList();
}


lib/src/json_to_firestore.dart
part of firestore_api_parser;

final firestoreReferenceRegex = RegExp(r'^projects\/.+\/databases\/\(default\)\/documents\/.+\/.+$');
final dateTimeRegex = RegExp(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(Z)?$'); // timezone is optional

@visibleForTesting
Map<String, dynamic> toFirestoreDocumentFormat({required Map<String, dynamic> jsonMap}) {
  final data = jsonMap.map((key, value) => MapEntry(key, encodeData(value)));

  return {'fields': data};
}

@visibleForTesting
Map<String, dynamic> encodeData(dynamic value) {
  if (value == null) return encodeNullData();

  if (value is String) {
    if (firestoreReferenceRegex.hasMatch(value)) return encodeReferenceData(value);

    final isValidDateTime = dateTimeRegex.hasMatch(value);
    if (isValidDateTime) return encodeTimestampData(value);

    return encodeStringData(value);
  }

  if (value is DateTime) return encodeTimestampData('${value.millisecondsSinceEpoch}');

  if (value is int) return encodeIntData(value);

  if (value is bool) return encodeBoolData(value);

  if (value is double) return encodeDoubleData(value);

  if (value is Map) {
    final isGeoPointData = value.containsKey('latitude') && value.containsKey('longitude') && value.entries.length == 2;
    if (isGeoPointData) return encodeGeoPointData(value);

    return encodeMapData(value);
  }

  if (value is List) return encodeListData(value);

  throw Exception("Type ${value.runtimeType} is not (for now) supported by our plugin.");
}

@visibleForTesting
Map<String, dynamic> encodeNullData() => {'nullValue': null};

@visibleForTesting
Map<String, dynamic> encodeBoolData(bool booleanValue) => {'booleanValue': booleanValue};

@visibleForTesting
Map<String, dynamic> encodeGeoPointData(Map geoPointValue) => {'geoPointValue': geoPointValue};

@visibleForTesting
Map<String, dynamic> encodeTimestampData(String timestampValue) => {'timestampValue': timestampValue};

@visibleForTesting
Map<String, dynamic> encodeReferenceData(String stringValue) => {'referenceValue': stringValue};

@visibleForTesting
Map<String, dynamic> encodeMapData(Map mapValue) {
  if (mapValue.isEmpty) return const {'mapValue': {}};

  final mapData = mapValue.map((key, value) => MapEntry(key, encodeData(value)));

  return {
    'mapValue': {'fields': mapData}
  };
}

@visibleForTesting
Map<String, dynamic> encodeListData(List arrayValue) {
  if (arrayValue.isEmpty) return const {'arrayValue': {}};

  final arrayData = arrayValue.map((value) => encodeData(value)).toList(growable: false);

  return {
    'arrayValue': {'values': arrayData}
  };
}

@visibleForTesting
Map<String, dynamic> encodeStringData(String stringValue) => {'stringValue': stringValue};

@visibleForTesting
Map<String, dynamic> encodeIntData(int integerValue) => {'integerValue': '$integerValue'};

@visibleForTesting
Map<String, dynamic> encodeDoubleData(double doubleValue) => {'doubleValue': doubleValue};


example/firestore_api_parser_example.dart
// Import firestore_api_parser
import 'package:firestore_api_parser/firestore_api_parser.dart';

import 'agent.dart';

void main() {
  final fsParser = FirestoreApiParser();

  final missions = <Map<String, String>>[
    {'2022': 'No Time To Die'},
    {'2015': 'Spectre'},
    {'2006': 'Casino Royale'},
  ];

  final names = {'lastname': 'BOND', 'firstname': 'James', 'initial': 'JB'};

  final bond = Agent(
    names: names,
    missions: missions,
    status: 'In service',
    round: 20.5,
    timestamp: DateTime.parse('2021-10-07T19:00:00Z'),
    storageRef: 'projects/my_great_project/databases/(default)/documents/USERS/[doc_id]',
    nullable: null,
    coordinates: {'latitude': -64, 'longitude': -86},
  );

  // Will convert [bond] to the format json used by Firestore format documents
  final firestoreJsonDoc = fsParser.parseJson(json: bond.toJson());

  // Will convert [firestoreJsonDoc] to a "normal" JSON format representation
  final json = fsParser.parseFirestoreDocument(documentJson: firestoreJsonDoc);

  // You can use this method bellow to parse & convert firestore collection to JSON
  // fsParser.parseCollection(firestoreCollection: firestoreCollection);
}


example/example.dart
import 'package:firestore_api_parser/firestore_api_parser.dart';

import 'agent.dart';

void main() {
  final names = {'lastname': 'BOND', 'firstname': 'James', 'initial': 'JB'};

  final bond = Agent(
    names: names,
    missions: [
      {'2022': 'No Time To Die'},
      {'2015': 'Spectre'},
      {'2006': 'Casino Royale'},
    ],
    status: 'In service',
    round: 20.5,
    timestamp: DateTime.parse('2021-10-07T19:00:00Z'),
    storageRef: 'projects/my_great_project/databases/(default)/documents/USERS/[doc_id]',
    nullable: null,
    coordinates: {'latitude': -64, 'longitude': -86},
  );

  // Will convert [bond] to the format json used by Firestore format documents
  final firestoreJsonDoc = FirestoreApiParser.convertToFirestoreDocument(json: bond.toJson());

  //using extension method
  // final firestoreJsonDoc = bond.toJson().toFirestoreJson;

  // Will convert [firestoreJsonDoc] to a "standard" JSON format representation
  final json = FirestoreApiParser.convertToStandardJson(doc: firestoreJsonDoc);

  //using extension method
  // final json =  firestoreJsonDoc.toStandardJson;

  // You can use this method bellow to parse & convert firestore collection to JSON
  // final document = FirestoreApiParser.parseCollection(collection: firestoreCollection);
}
